### PR TITLE
Followon to #14559

### PR DESCRIPTION
**Pull request by mheon**
Taking over #14559 from @vrothberg and trying to get CI to pass.

No changes (besides a rebase) yes, I just want to verify we have the same symptoms.

```release-note
Fix a bug in the wait logic to enable Podman on Gitlab runner.
```

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
The commit message saying "gitlab:" as the prefix makes no sense. Can you please drop it?

**Comment by mheon** [comment]
Got inside a CI VM to test. I was only able to reproduce an actual timeout once, and it was on `podman rm -t 0`, not `podman kill`. The cause was one of more stuck cleanup processes holding container locks. Still unclear as to why they were stuck, and how that could affect `podman kill` as it does in the CI runs (cleanup process should only be running *after* kill?).

**Comment by mheon** [comment]
ID'd some potential issues in the DB. Have not had a chance to test, but I'll let CI bake overnight.

**Comment by mheon** [comment]
Hot damn it's actually passing (minus a flake).

I'll clean this up in the morning, then we can get final review + merge done.

File: libpod/boltdb_state.go [hunk 1]
```diff
@@ -1341,6 +1391,175 @@ func (s *BoltState) GetContainerConfig(id string) (*ContainerConfig, error) {
...
+
+		if err := result.UnmarshalText(rawTimeStamp); err != nil {
+			return fmt.Errorf("converting raw time stamp %v of container %s from DB: %w", rawTimeStamp, id, err)
+		}
+
+		return nil
+	})
+}
+
+// PrunExitCodes removes exit codes older than 5 minutes.
```
Comment (giuseppe) [inline libpod/boltdb_state.go:1494]: typo in the comment

File: libpod/boltdb_state.go [hunk 2]
```diff
@@ -1341,6 +1391,175 @@ func (s *BoltState) GetContainerConfig(id string) (*ContainerConfig, error) {
...
+				if err := exitCodeBucket.Delete(rawID); err != nil {
+					if finalErr != nil {
+						logrus.Error(finalErr)
+					}
+					finalErr = fmt.Errorf("removing exit code of container %s from DB: %w", id, err)
+				}
+				if err := timeStampBucket.Delete(rawID); err != nil {
+					if finalErr != nil {
+						logrus.Error(finalErr)
+					}
```
Comment (giuseppe) [inline libpod/boltdb_state.go:1549]: should we update `finalErr = err` to avoid reporting twice the same error?

**Review by giuseppe — COMMENTED** [review]
two non-blocking nits, otherwise LGTM

**Comment by mheon** [comment]
Force-pushed, should be ready now.

**Comment by mheon** [comment]
@baude @Luap99 @rhatdan PTAL

**Comment by mheon** [comment]
Oops, fixed

File: libpod/events.go [hunk 3]
```diff
@@ -151,6 +151,9 @@ func (r *Runtime) GetEvents(ctx context.Context, filters []string) ([]*events.Ev
 // GetLastContainerEvent takes a container name or ID and an event status and returns
 // the last occurrence of the container event
 func (r *Runtime) GetLastContainerEvent(ctx context.Context, nameOrID string, containerEvent events.Status) (*events.Event, error) {
+	// FIXME: events should be read in reverse order!
+	// https://github.com/containers/podman/issues/14579
```
Comment (Luap99) [inline libpod/events.go:155]: This issue is closed, either remove this comment or if we really need it reopen the issue.

Comment (mheon) [reply libpod/events.go:155]: I'm going to repoen, I think it's a worthwhile change - but probably low priority.

**Comment by mheon** [comment]
Ready to merge

**Review by vrothberg — COMMENTED** [review]
LGTM
Thanks for kicking it over the finish line, @mheon. Great work!

**Comment by mheon** [comment]
@baude @Luap99 @rhatdan PTAL and merge

**Comment by vrothberg** [comment]
@giuseppe PTAL

**Review by giuseppe — APPROVED** [review]
/lgtm

**Comment by runiq** [comment]
As an outsider: Good job, everyone! :) Looking forward to (eventually) using Podman in Gitlab.

**Pull request merged**